libglusterfs/src/mem-pool.c

```diff
@@ -363,17 +362,14 @@ __gf_free(void *free_ptr)
         GF_ASSERT(GF_MEM_TRAILER_MAGIC == __gf_mem_trailer_read(trailer));
     }
 
-    num_allocs = GF_ATOMIC_DEC(mem_acct->rec[header->type].num_allocs);
+    GF_ATOMIC_DEC(mem_acct->rec[header->type].num_allocs);
 #ifdef DEBUG
     LOCK(&mem_acct->rec[header->type].lock);
     {
         list_del(&header->acct_list);
     }
     UNLOCK(&mem_acct->rec[header->type].lock);
 #endif
-    if (!num_allocs) {
```

Comment:
I think @mohit84 mentioned the CLI might be such an example - in which case, I don't think we really care if it leaks. I do wonder if we have other cases though. Code coverage seem to hint it's a rare case:
```
760 575550743 :     if (GF_ATOMIC_DEC(mem_acct->refcnt) == 0) {]()
761             : #ifdef DEBUG]()
762    18850267 :         for (i = 0; i < mem_acct->num_types; i++) {]()
763    18680910 :             LOCK_DESTROY(&(mem_acct->rec[i].lock));]()
764             :         }]()
765             : #endif]()
766      169357 :         FREE(mem_acct);]()                <--- out of 575550743 calls to the 'if' statement, only 169357 landed here.
```

Reply:
If we remove it it will become a leak of mem_acct for every xlator. For single brick it does not matter but In case of brick_mux you can easily observe it.

Reply:
> I think @mohit84 mentioned the CLI might be such an example - in which case, I don't think we really care if it leaks. I do wonder if we have other cases though. Code coverage seem to hint it's a rare case:
> 
> ```
> 760 575550743 :     if (GF_ATOMIC_DEC(mem_acct->refcnt) == 0) {]()
> 761             : #ifdef DEBUG]()
> 762    18850267 :         for (i = 0; i < mem_acct->num_types; i++) {]()
> 763    18680910 :             LOCK_DESTROY(&(mem_acct->rec[i].lock));]()
> 764             :         }]()
> 765             : #endif]()
> 766      169357 :         FREE(mem_acct);]()                <--- out of 575550743 calls to the 'if' statement, only 169357 landed here.
> ```
> 
> (not that I fully understand the code coverage here - how come the lines in the DEBUG, got some, more than line 766? Am I missing some optimization? Thought code coverage was with debug and no optimizations...)

Lines 762 and 763 are executed several times because of the loop for each free of the `mem_acct`. In fact, #762 = #763 + #766, meaning that the loop executed one more time (the last check to identify the end of the loop) than the body of the loop for each destruction of a `mem_acct`.